pkg/datapath/loader/base.go

```diff
@@ -357,13 +357,15 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
 	if option.Config.IPAM == ipamOption.IPAMENI {
 		// For the ENI ipam mode on EKS, this will be the interface that
 		// the router (cilium_host) IP is associated to.
-		if info := node.GetRouterInfo(); info != nil {
-			mac := info.GetMac()
-			iface, err := linuxrouting.RetrieveIfaceNameFromMAC(mac.String())
-			if err != nil {
-				log.WithError(err).WithField("mac", mac).Fatal("Failed to set encrypt interface in the ENI ipam mode")
+		if option.Config.EncryptInterface == "" {
```

Comment:
How does this work if user specifies a different interface from the one `cilium_host` is associated to? Encrypted traffic will arrive at the destination node on the primary ENI interface, but if the `bpf_network` program to decrypt traffic is attached to a different interface selected by users, then packets won't get decrypted? Are we expecting users to provide the correct interface by looking at the `cilium_host` ENI association?

Reply:
In this series with the subnet mode, commit `cilium: encryption, ENI with subnet encryption needs rules for link IP` (currently bc534be5e29343f4bd34a7aaa7b883df37ecf8ed) is proposing to switch away from reliably using the cilium_host IP for the local encryption IP to instead pick an IP from the `option.Config.EncryptInterface`. So there won't be a dependency on the cilium_host IP any more for this configuration.

Reply:
I've got the impression a few times from discussions with John that in this mode, the IP used for encryption isn't necessarily the cilium_host IP but I don't really understand where it comes from. @jrfastab is that true? Is there something about subnet mode that causes different IPs to be used for the encryption?

That said, this particular patch is basically just saying "if the user explicitly configures the encrypt interface, do what they say, even if it's wrong because if they're explicitly setting this configuration, they know what they're doing". In general we should not automatically override some explicit user configuration.

Reply:
>Are we expecting users to provide the correct interface by looking at the cilium_host ENI association?

I have a docs PR in progress where I mention that the `option.Config.EncryptInterface` config won't take effect in case of the IPAM ENI mode.

Reply:
Problem is at least in EKS (k8s 1.19) the detected interface is wrong, eth1 instead of eth0 in my case.  I haven't investigated yet why its the wrong interface yet, but merely noting we shouldn't ignore user input. If a user pushes down an explicit interface then we should honor it. If we want a WARN or something if they don't match that would be nice I guess.

In subnet mode, required in modes with a global IP pool, we set src IP using the IPV4_ENCRYPT_IFACE chosen here,
```
        if iface := option.Config.EncryptInterface; iface != "" {
                link, err := netlink.LinkByName(iface)
                if err == nil {
                        cDefinesMap["ENCRYPT_IFACE"] = fmt.Sprintf("%d", link.Attrs().Index)

                        addr, err := netlink.AddrList(link, netlink.FAMILY_V4)
                        if err != nil {
                                return err
                        }
                        if len(addr) == 0 {
                                return fmt.Errorf("no IPv4 addresses available in encrypt interface %q", iface)
                        }
                        a := byteorder.HostSliceToNetwork(addr[0].IPNet.IP, reflect.Uint32).(uint32)
                        cDefinesMap["IPV4_ENCRYPT_IFACE"] = fmt.Sprintf("%d", a)
                }
        }
```
Why not cilum_host in this mode? The initial idea here was to avoid leaking podIPs and CiliumIPs into the network and only use networkIPs.

Reply:
Huh, weird, so that logic you point to above doesn't have any link to the logic that this comment thread is looking at. Seems like if you don't configure `option.Config.EncryptInterface` then `ENCRYPT_IFACE` and  `IPV4_ENCRYPT_IFACE` won't get set. Might explain why the autodetection isn't lining up with the other parts of the datapath for your setup.

Reply:
>In subnet mode, required in modes with a global IP pool, we set src IP using the IPV4_ENCRYPT_IFACE chosen here,

To keep the scope of the original fix for the IPAN ENI issue limited, I left this as a follow-up with a note in the original PR #https://github.com/cilium/cilium/pull/14924. :) In the code snippet you've added above, it also mandates that the interface an an IP address. That's not the case for the IPAN ENI mode.

Reply:
Agree I think it needs a follow up. What do you mean by 'thats not the case for the IPAM ENI mode"? Do you mean in AWS-ENI? My eth0 has an IP address in that configuration,
```
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc mq state UP group default qlen 1000
    link/ether 02:52:43:a6:f3:43 brd ff:ff:ff:ff:ff:ff
    inet 192.168.90.141/19 brd 192.168.95.255 scope global dynamic eth0
       valid_lft 3056sec preferred_lft 3056sec
    inet6 fe80::52:43ff:fea6:f343/64 scope link 
       valid_lft forever preferred_lft forever
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9001 qdisc mq state UP group default qlen 1000
    link/ether 02:67:c8:f5:87:87 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::67:c8ff:fef5:8787/64 scope link 
       valid_lft forever preferred_lft forever
```
eth1 however does not.

Reply:
Yes, that's exactly what I was referring to. On EKS, eth1 will be auto-detected.

Reply:
Logging a warning sounds like a decent middle ground.